service/controller/v11/resource/endpoint/current.go

```diff
@@ -0,0 +1,76 @@
+package endpoint
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/finalizerskeptcontext"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/kvm-operator/service/controller/v11/key"
+)
+
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	pod, err := toPod(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "looking for annotations on pod")
+
+	_, serviceName, err := getAnnotations(*pod, IPAnnotation, ServiceAnnotation)
+	if IsMissingAnnotationError(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "annotation is missing on pod")
+		resourcecanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for pod")
+
+		return nil, nil
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	if key.IsPodDeleted(pod) {
+		a := pod.GetAnnotations()
+		if a == nil {
+			return nil, microerror.Mask(missingAnnotationError)
+		}
+		v, ok := a[key.AnnotationPodDrained]
+		if !ok {
+			return nil, microerror.Mask(missingAnnotationError)
+		}
+		if v == "False" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "cannot finish deletion of pod due to undrained status")
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil, nil
+		}
+	}
+
+	currentEndpoint := Endpoint{
+		IPs:              []string{},
+		ServiceName:      serviceName,
+		ServiceNamespace: pod.GetNamespace(),
+	}
```

Comment:
Why do we have separate type for state instead of `*corev1.Endpoints`?

Reply:
Copy/paste. Don't care. Not happy with this but this is out of the scope of this PR. Ask Marcel maybe. We can iterate here later on.

Reply:
Ack.